### PR TITLE
Implement generic fromBin|Oct|Hex in strutils

### PR DESCRIPTION
**Pull request by avitkauskas**
Implement all families of proc like `parseBinInt8`, `parseBinInt16`, ..., `parseBinUint64` (and Oct|Hex) in strutils returning their respective types of integer.

### CODE REVIEW

**Comment by kaushalmodi** [comment]
Is it possible to design generics for this? I don't like the idea of hardcoding types (int, uint, etc) in proc names.

**Comment by kaushalmodi** [comment]
There was an awesome effort put in by @timotheecour a while back: https://github.com/nim-lang/Nim/pull/10346, but unfortunately they PR got closed.

**Comment by avitkauskas** [comment]
@kaushalmodi I think this could be easily done by:
```nim
proc parseFromBin*[T](v: var T, s: string): T =
     let p = parseutils.parseBin(s, result)
     if p != s.len or p == 0:
       raise newException(ValueError, "invalid binary integer: " & s)

var bin: int8
bin = bin.parseFromBin("0b11011")
```

... or something. But there is already the proc `parseBinInt` in stdlib, so I wanted just to expand this to other types. Actually `parsutils.parseBin` is now generic and almost the same, but without the Exception guard.

**Comment by kaushalmodi** [comment]
> But there is already the proc parseBinInt in stdlib

Then may be that should be deprecated? Asking @Araq to weigh in on this.

**Comment by avitkauskas** [comment]
Mistery - koch changes oct literal when generating doc examples!

my code in strutils.nim:
```nim
runnableExamples:
    let a = "0o_123_456_777"
    doAssert a.parseOctInt8() == 0o377'i8
```

./koch doc generates in example file:
```nim
block:
  let a = "0o_123_456_777"
  doAssert a.parseOctInt8() == 0o777'i8  # <--- see this!
```

and fails with the error:
```
strutils_examples29.nim(9, 32) Error: number out of range: '0o777'i8'
```

If I try `0o277'i8` it fails with `0o677'i8`.
Is that my problem, or koch?

**Comment by mratsim** [comment]
0o777 is 0x1FF in hexadecimal, a byte can represent from 0x00 to 0xFF.

**Comment by avitkauskas** [comment]
The thing is - I put in runableExamples `0o377'i8` (and that's 0xFF, correct), but `koch doc` when generating examples changes this to `0o777'i8` in the generated examples file, and then of course tests fail with the error message `Error: number out of range: '0o777'i8'`.

But how can be that this change happens while extracting my code into separate examples file?
All other lines arround this one are transfered to example file verbatim without any changes.

**Comment by Araq** [comment]
> But how can it be that this change happens while extracting my code into separate examples file?

Well the "extraction" is ... er... special. :-)

**Comment by data-man** [comment]
parseX**Int**X in the docs, but parseX**int**X by fact.
Isn't good especially for autocompletion.

**Comment by avitkauskas** [comment]
@data-man Yes, you are right, there would be a problem for autocompletion as you say.
So, maybe if would be much easier and better then to introduce procs like:

```nim
proc fromBin*[T: SomeInteger](v: var T, s: string) =
     let p = parseutils.parseBin(s, v)
     if p != s.len or p == 0:
       raise newException(ValueError, "invalid binary integer: " & s)

var x: int8
x.fromBin("0b11011")
```

Especially when there are already procs `toBin` etc. in `strutils`.

And then we could probably deprecate `parseBinInt` if we get the agreement on that.

**Comment by data-man** [comment]
Personally I'd prefer universal ```parseInt``` with ```base``` parameter. And deprecate others.
What if someone wants to add a ```parseTernaryInt``` and family tomorrow? Or/and ```parseSeptenaryInt``` and family?
strutils is already big enough.

**Comment by mratsim** [comment]
They had this parseTernaryInt in their own library just like I did here: https://github.com/status-im/nim-stint/blob/6853ebe97c21426952ce64c2df6278514797dc3f/stint/io.nim#L116-L117

```Nim
func parse*[bits: static[int]](input: string, T: typedesc[Stuint[bits]], radix: static[uint8] = 10): T =
  ## Parse a string and store the result in a Stuint[bits].

func parse*[bits: static[int]](input: string, T: typedesc[Stint[bits]], radix: static[int8] = 10): T =
  ## Parse a string and store the result in a Stint[bits] or Stuint[bits].
```

File: lib/pure/strutils.nim [hunk 1]
```diff
@@ -1068,44 +1068,145 @@ proc parseFloat*(s: string): float {.noSideEffect, procvar,
 
 proc parseBinInt*(s: string): int {.noSideEffect, procvar,
   rtl, extern: "nsuParseBinInt".} =
-  ## Parses a binary integer value contained in `s`.
+  ## Parses a binary integer value contained in `s` and returns `int`.
```
Comment (Araq) [inline lib/pure/strutils.nim]: "And returns int"? Do we need to assume nobody can read Nim type declarations?

**Comment by Araq** [comment]
I appreciate the effort but this should use generics indeed. You can then deprecate the non-generic variant or simply leave it as it is.

**Comment by avitkauskas** [comment]
@Araq When talking about generics, what variant would you prefer:

```nim
proc fromBin*[T: SomeInteger](x: var T, s: string) =
  let p = parseutils.parseBin(s, x)
  if p != s.len or p == 0:
    raise newException(ValueError, "invalid binary integer: " & s)

var x: int8
x.fromBin("0b11011")
```

or

```nim
proc fromBin*[T: SomeInteger](s: string): T =
  let p = parseBin(s, result)
  if p != s.len or p == 0:
    raise newException(ValueError, "invalid binary integer: " & s)

var x: int8
x = fromBin[int8]("0b11011")
# by the way, x = "0b11011".fromBin[int8] would not work: cannot instantiate: 'T' and other errors
```

**Comment by Araq** [comment]
This variant: 

```nim

proc fromBin*[T: SomeInteger](s: string): T =
  let p = parseutils.parseBin(s, result)
  if p != s.len or p == 0:
    raise newException(ValueError, "invalid binary integer: " & s)

var x: int8
x = fromBin[int8]("0b11011")

# by the way, x = "0b11011".fromBin[:int8] would work perfectly fine.
```

**Comment by mratsim** [comment]
I feel like we're back to typedesc vs generics API design here with a sprinkle of backward compatibility concerns (https://github.com/nim-lang/RFCs/issues/40)

**Comment by Araq** [comment]
> I feel like we're back to typedesc vs generics API design here with a sprinkle of backward compatibility concerns (nim-lang/RFCs#40)

The stdlib uses ``[T]`` and so that's what should be used. I can find an RFC for everything that could be used to delay further progress.

**Comment by avitkauskas** [comment]
runnableExamples tests will not pass until #11131 is fixed.

**Comment by avitkauskas** [comment]
What should I do about that?
```
Hint: test_vectors [Processing]
1414test_vectors.nim(142, 27) Error: ambiguous call; both utils.fromHex(a: string) [declared in ..\..\..\..\..\Users\appveyor\.nimble\pkgs\nimcrypto-0.3.9\nimcrypto\utils.nim(158, 6)] and strutils.fromHex(s: string) [declared in ..\..\..\lib\pure\strutils.nim(1059, 6)] match for: (string)
```

**Comment by narimiran** [comment]
> runnableExamples tests will not pass until #11131 is fixed.

This is fixed now.

> What should I do about that?

Where does `test_vectors.nim` come from? I didn't find it in Nim repo, is it some Nimble package?

**Comment by avitkauskas** [comment]
@narimiran, it's in `nimcrypto-0.3.9` package as I understand.

**Comment by narimiran** [comment]
> it's in nimcrypto-0.3.9 package as I understand.

Ah, I found it. It is `blscurve` package that is failing, and here is [test_vectors.nim](https://github.com/status-im/nim-blscurve/blob/master/tests/test_vectors.nim)

```
test_vectors.nim(142, 27) Error: ambiguous call; both utils.fromHex(a: string) [declared in nimcrypto\utils.nim(158, 6)] and strutils.fromHex(s: string) [declared in lib\pure\strutils.nim(1059, 6)] match for: (string)
````

This can then be fixed by patching the mentioned package by having this line:
```nim
import strutils except fromHex
```

**Comment by mratsim** [comment]
Merged https://github.com/status-im/nim-blscurve/pull/21 and retriggering CI. Can we put Appveyor under a Nim organisation. I can restart Travis builds but not Appveyor's.

**Comment by sophisticate364** [comment]
Why is it merging my account ??and they gain access thru making URLs
________________________________

**Comment by avitkauskas** [comment]
This is ready for review.

**Pull request merged**